St abstraction

EnguinandP opened the pull request:
Worked on the highest two priority issues so far. I need more clarification on abstracting away the random state "st" parameter.

Francille committed:
Parametrized generators

Francille committed:
Abstracted away random state st

Not 100% sure this is right, I'll ask for clarification in the morning. I'm using a ref, but not with option (I'm not sure if that's significant)

EnguinandP commented:
Uh oh, it looks like I didn't rebase with main properly, and I merged instead. Sorry if it looks confusing!

Francille committed:
Added int list gen with variable size

Also fixed "postcondition" and made minor changes for csv files

Pat-Lafon commented:
I was going to say the history is fine, except now the branch has a bunch of conflicts apparently. In general, I have set my pull strategy to rebase in git to try to avoid this. Maybe you can fix it via a rebase?

Pat-Lafon commented on bin/combinators.ml:
```diff
@@ -1,14 +1,75 @@
+(* higher order programming *)
+(* uses QCheck.Gen.nat for random size of list *)
+(* make still expects random state as parameter, "_" gets rid of it*)
+let arb_builer f = QCheck.make (fun _ -> f (QCheck.Gen.nat !st) )
```
So what we want to do here I think is take the state that QCheck will give us and assign it to `st`.

Pat-Lafon replied:
I think this builder needs a second thought, I don't want to overfit on there being this nat first argument, and then have the rest of the builder figure out the rest of the arguments.

EnguinandP replied:
Hmm the state from Qcheck is supplied in `Qcheck.run_tests_make`, which has the option of being passed in as a parameter. Is the idea to instead of letting `Qcheck.run_tests_make` apply the random state, apply the random state in the individual generators? I understand that part of the goal to remove the `st` parameter because your generators in cobb don't have them, right?

Pat-Lafon replied:
Yeah, in some sense, the fact that the generators use some `st` to generate deterministic values is an implementation detail that I want to hide away(and not relevant in the synthesis process). So I am trying to align what you are doing with how those generators are written so at some point we can have both. Your specialized generators and whatever I can get Cobb to synthesize so we can compare them

Pat-Lafon replied:
My mental model is that QCheck is supplying to use some random state to use. Maybe this is not correct. If it is, I want to continue using this random state in our generator but I want to hide it away so that we can just as easily plug in a synthesized generator as we plug in your specialized generator

EnguinandP replied:
How do the cobb generators maintain deterministic values then?

EnguinandP replied:
Ok, so I know that `Qcheck.run_tests_make` can either take in a seed to create a random state or take in a random state itself (if neither, it will create a random, random state). Would it be enough to
1. create a random state in main
2. pass it into `Qcheck.run_tests_make`
2. remove the `st` parameters from my generators and pass in the pre-initialized random state?

Pat-Lafon replied:
The Cobb generators are currently processed symbolically(via type-checking), the goal of some of your work is to eventually get them to execute and compare them against some alternative generators in an numerical sense.

Pat-Lafon replied:
To be clear, I'm not talking about messing with the seed I think. Yes, QCheck internally turns that seed into a random state, but then it hands it off to the generator to create a value from that state. That is a part of the type signature of gen `type 'a t = Random.State.t -> 'a `. Basically instead of our generators looking something like that I want to break it into two parts. A `Random.State.t -> ()` step that hides the random state by assigning it to a global ref, and a `() -> 'a` step which is sorta the signature of our generator(like an int_gen generator). Internally, `int_gen` can access this random state to create a value

Pat-Lafon commented on bin/main.ml:
```diff
@@ -13,23 +13,25 @@ let rec is_duplicate = function
   ~count:20000
   ~name
-   (pair (int) (list int)) (fun (n, l) ->
+   (pair (int) (gen_type)) (fun (n, l) ->
```
Not all functions will need this pair of int/gen_type. We can instead replace this with just gen_type and have each benchmark fully supply it's required generator

Pat-Lafon replied:
Notice here how we are unpacking the pair of inputs given to us to run the precondition against

Pat-Lafon commented on bin/main.ml:
```diff
@@ -13,23 +13,25 @@ let rec is_duplicate = function
-let precondition_frequency_dup = precondition_frequency (fun _ l -> is_duplicate l) "is_duplicate"
-let precondition_frequency_unique = precondition_frequency (fun _ l -> is_unique l) "is_unique"
+let precondition_frequency_size = precondition_frequency is_sized "is_sized" default_gen
+let precondition_frequency_sort = precondition_frequency (fun _ l -> is_sorted l) "is_sorted" default_gen
```
Then we shouldn't need these lambda's that just ignore the first argument

Pat-Lafon commented on scripts/output.py:
```diff
@@ -15,14 +16,23 @@ def read_input():
             # print(extracted_values)
             extracted_values.insert(0, extracted_values[3])
             extracted_values.pop()
+            extracted_values.insert(1, extracted_values[2])
+            del extracted_values[3]
```
Hmm, low priority but could we just use a dictionary for all of this? Hopefully less error prone and then we should never have issues with column alignment like we have previously

Pat-Lafon replied:
Moved into it's own issue so I will resolve this

Pat-Lafon commented on bin/combinators.ml:
```diff
@@ -1,14 +1,75 @@
+    QCheck.Gen.int !st :: int_list_size_gen (s - 1)
+
+(* int list gen of size s or less*)
+let int_list_variable_size_gen s =
+  let s' = Random.State.int !st (s + 1) in 
+    int_list_size_gen s'
```
Hmm, nice idea here. Can I challenge you to make this a part of int_list_size_gen instead?

EnguinandP replied:
I can try. We still need to keep `int_list_size_gen` as itself so that we have a default list gen (that doesn't match any preconditions), right?

Pat-Lafon replied:
We can always reach for the built in generator like `(list int)` or something that QCheck supplied as a reasonable default if you don't want two implementations floating around

EnguinandP replied:
There's also a problem that I'm trying to resolve with `int_list_variable_size_gen`. For `int_list_variable_size_gen` to work, it has to have the same size parameter as the precondition. However, there isn't really a straighforward way to pass in a size to both the precondition and the generator.

`let precondition_frequency_size = 
  let s = QCheck.Gen.int !Combinators.st in
    precondition_frequency (is_sized s) "is_sized" (Combinators.int_list_variable_size_gen s)`

Like there really isn't a way to get a var to Combinators.int_list_variable_size_gen since it's an arbitrary type and not a function.

Pat-Lafon replied:
Yeah, the problem your having is because you think the generator only produces one type. It doesn't. Think of the `pair` combinator. Ideally we are trying to figure out the right set of building blocks that compose nicely.

I would expect you to bundle each of your generated inputs together, and then unbundle them to call the function(ok sure, it's maybe an unnecessary level of indirection but we need to move away from hardcoding these inputs somehow)

Pat-Lafon commented:
I would like you to create a set of simple components/functions to use to construct generators that will hide away the `st`. They should be called `int_gen ()`, `bool_gen ()`, and `list_gen ()`. They each take no arguments and produce a random value of the given type. There will probably be more components to add after this to align with the components I use in synthesis.

EnguinandP commented:
> I was going to say the history is fine, except now the branch has a bunch of conflicts apparently. In general, I have set my pull strategy to rebase in git to try to avoid this. Maybe you can fix it via a rebase?

Ok, I just rebased the branch from last week onto this branch. It looks like I have to make a new PR for history to line up?

Pat-Lafon commented:
As in, can you rebase off of this repositories main branch? If you have this setup as your remote repository, it may just be `git rebase remote/main` or `upstream/main` or something like that. (might need to git fetch first)

EnguinandP commented:
Yeah, I tried that too and `git push --force`. It looks alright in my PR preview (says merging 3 commits into your main)?

Pat-Lafon commented:
Here I see 12 commits and some of them I think you have already merged like "Created a new folder for .csv"

EnguinandP commented:
Yeah, the problem was that I had used merge with main/last week's branch instead of rebase. It's now rebased properly in my repo, but it doesn't appear in our PR. Do you know what I could do to fix this?

Pat-Lafon commented:
> Yeah, the problem was that I had used merge with main/last week's branch instead of rebase. It's now rebased properly in my repo, but it doesn't appear in our PR. Do you know what I could do to fix this?

When you say rebased in your repo, are you rebasing against your main branch or mine?

EnguinandP commented:
Oh! I was not rebasing against yours. Give me a second

EnguinandP commented:
Sorry, that was my unfamiliarity with working with remotes, my bad!

Pat-Lafon commented:
All good, it's kinda tricky anyways

Pat-Lafon commented on bin/combinators.ml:
```diff
@@ -3,55 +3,71 @@
-let int_list_unique = list int_list_unique_gen
+let int_list = arb_builer int_list_size_gen
+let int_list_variable_size = arb_builer int_list_variable_size_gen
+let int_list_sorted = arb_builer (int_list_sorted_gen 0)
```
I'm realizing I'm not sure why sorted lists are hardcoded to zero here

EnguinandP replied:
This is a little outdated now, but `int_list_sorted` was a recursive function that cons a new element based on the previous element. `prev` is passed as a parameter and it starts at 0 (since it’s ascending order). I should’ve use an aux in that case for readability, but now it does.

Pat-Lafon replied:
Instead of always starting from zero, why not start from some arbitrary integer that we can pass in by a parameter?

Francille committed:
Created int_gen, bool_gen

Attempted adding list_gen, removing nat gen from arb_builder, and improving int_list_variable_size. They need more consideration

Pat-Lafon commented on bin/combinators.ml:
```diff
@@ -7,6 +7,9 @@
 intentions are, will ask for clarification on Monday *)
 let st = ref (Random.State.make [| 0 |])
 
+let int_gen = QCheck.Gen.int !st
```
Notice the difference between `let int_gen = ...` and `let int_gen () = ...`

EnguinandP replied:
Oh, I don’t think I saw this comment (I still can’t find it either?)

Pat-Lafon replied:
Ooof

EnguinandP commented:
> I would like you to create a set of simple components/functions to use to construct generators that will hide away the `st`. They should be called `int_gen ()`, `bool_gen ()`, and `list_gen ()`. They each take no arguments and produce a random value of the given type. There will probably be more components to add after this to align with the components I use in synthesis.

There's something I'm confused about in this part. With the old generators that take parameter `st`, the Qcheck runners supply the the random start. If we take out the parameter `st`, then we are supplying the random state ourselves.  Would that still be semantically consistent?

Francille committed:
Expanded the state ref

Confused about abstracting away st with current implementation, it seems to cause inconsistent behavior

EnguinandP commented on bin/combinators.ml:
```diff
@@ -1,23 +1,27 @@
+| Some s -> s
+| None -> set_st
+
+let int_gen = QCheck.Gen.int (get_st st')
```
I've replaced the previous cases of `Qcheck.Gen.int st` with `int_gen`. `int_gen` doesn't take in any random states, instead it uses `ref st'` where `st` then gets initialized in main using `st_set`. I have doubts whether this is correct however, since running Qcheck with these new generators, gives me not only different results, but really confusing results. I don't if implementing these this way, it's initialized the random numbers in the generators (and it's not regenerating new ones each time).

Pat-Lafon replied:
I'm a little confused with this level of indirection. You should only need one st. You can either initialize it to some random seed or leave it unintialized. Then it should be `set_st st = ...` which we call with the st from the runner and then `get_st () = !st` or `get_st () = Option.value !st` (I just guess what the right Option.method was, it might be something different like get)

EnguinandP commented on csv/unique_int_list_gen.csv:
```diff
@@ -1,5 +1,5 @@
-is_sorted,20200,3905,0.7s
-is_duplicate,20200,3187,0.8s
-is_unique,20000,20000,1.4s
+Precondition Name,Pass,Total,Time
```
generated error  fail  pass / total     time test name
[✓] 20200     0     0 10152 / 20000     0.3s is_sized
[✓] 20000     0     0 20000 / 20000     0.3s is_sorted
[✓] 20000     0     0 20000 / 20000     0.3s is_duplicate
[✓] 20000     0     0 20000 / 20000     0.2s is_unique

This is an example of the results of using `int_list_unique` generator with the new changes, and as far as I can tell, everything else about the generator and the preconditions are the same.

Pat-Lafon replied:
So is the problem you are saying that `int_list_unique` generates completely valid is_duplicate lists?

Pat-Lafon replied:
See my other comment. is_duplicate and is_unique agree when you are only checking singleton/duplicate lists

Pat-Lafon replied:
I'm assuming this is fixed now when new csv's are pushed?

EnguinandP replied:
Yes, this is resolved (sorry, I missed this), the new csv files are now correct

Pat-Lafon commented:
> > I would like you to create a set of simple components/functions to use to construct generators that will hide away the `st`. They should be called `int_gen ()`, `bool_gen ()`, and `list_gen ()`. They each take no arguments and produce a random value of the given type. There will probably be more components to add after this to align with the components I use in synthesis.
> 
> There's something I'm confused about in this part. With the old generators that take parameter `st`, the Qcheck runners supply the the random start. If we take out the parameter `st`, then we are supplying the random state ourselves. Would that still be semantically consistent?

That is what we need to add in. Instead of assigning our own `st` from a seed, we want to use the QCheck runner's random `st`

Pat-Lafon commented on bin/combinators.ml:
Ah! @Francille see my comment on the difference between `int_gen` and `int_gen ()` here. My guess is that this is only generating singleton/empty lists? That is when is_duplicate and is_unique agree no?

EnguinandP replied:
Oh man, I missed this one too!

Pat-Lafon replied:
Yeah, maybe they were only on my end since I hadn't hit my overall review... On my end it had these in the pr... whoops, my bad

Francille committed:
Fixed st abtstraction

Fixes #12, Fixes #10, Fixes #9, Fixes #8

EnguinandP commented:
> > > I would like you to create a set of simple components/functions to use to construct generators that will hide away the `st`. They should be called `int_gen ()`, `bool_gen ()`, and `list_gen ()`. They each take no arguments and produce a random value of the given type. There will probably be more components to add after this to align with the components I use in synthesis.
> > 
> > 
> > There's something I'm confused about in this part. With the old generators that take parameter `st`, the Qcheck runners supply the the random start. If we take out the parameter `st`, then we are supplying the random state ourselves. Would that still be semantically consistent?
> 
> That is what we need to add in. Instead of assigning our own `st` from a seed, we want to use the QCheck runner's random `st`

That is a much more obvious solution, I just fixed it.

Pat-Lafon commented:
> > > > I would like you to create a set of simple components/functions to use to construct generators that will hide away the `st`. They should be called `int_gen ()`, `bool_gen ()`, and `list_gen ()`. They each take no arguments and produce a random value of the given type. There will probably be more components to add after this to align with the components I use in synthesis.
> > > 
> > > 
> > > There's something I'm confused about in this part. With the old generators that take parameter `st`, the Qcheck runners supply the the random start. If we take out the parameter `st`, then we are supplying the random state ourselves. Would that still be semantically consistent?
> > 
> > 
> > That is what we need to add in. Instead of assigning our own `st` from a seed, we want to use the QCheck runner's random `st`
> 
> That is a much more obvious solution, I just fixed it.

Ah, I didn't know that we can arbitrarily request the random state with `QCheck_runner.random_state`. Good find

EnguinandP commented:
> > > > > I would like you to create a set of simple components/functions to use to construct generators that will hide away the `st`. They should be called `int_gen ()`, `bool_gen ()`, and `list_gen ()`. They each take no arguments and produce a random value of the given type. There will probably be more components to add after this to align with the components I use in synthesis.
> > > > 
> > > > 
> > > > There's something I'm confused about in this part. With the old generators that take parameter `st`, the Qcheck runners supply the the random start. If we take out the parameter `st`, then we are supplying the random state ourselves. Would that still be semantically consistent?
> > > 
> > > 
> > > That is what we need to add in. Instead of assigning our own `st` from a seed, we want to use the QCheck runner's random `st`
> > 
> > 
> > That is a much more obvious solution, I just fixed it.
> 
> Ah, I didn't know that we can arbitrarily request the random state with `QCheck_runner.random_state`. Good find

Yeah, I saw it couple times while struggling with `st` but didn't think much of it until your comment lol

EnguinandP commented on bin/combinators.ml:
```diff
@@ -1,59 +1,83 @@
-  else
-    QCheck.Gen.int st :: int_list_size_gen (s - 1) st
+(* default int list gen with size s*)
+let int_list_gen () = 
+  let size = nat_gen () in
+  let rec aux s =
+    if s <= 0 then
+      []
+    else
+      int_gen () :: aux (s - 1) in
+  aux size
```
Should I change the function signatures of all the generators to be like `let int_list_gen size`? Is there anything else?

Pat-Lafon replied:
Maybe this is a good time to look at the the Cobb synthesizers and get a feel for what I'm currently working with. Note that there are many ways to write some of these generators... but these are the ones I got from the previous work I'm building upon so we should stick close to those.

https://github.com/Pat-Lafon/underapproximation_type/blob/0c760d66a814827fae1b5b250b24e310891127a6/data/validation/sizedlist/prog.ml

Pat-Lafon replied:
Note that sortedlist and some of the other generators do take additional arguments other than just a size argument

EnguinandP replied:
Yeah I saw that, I assume that means using a tuple instead of a pair? (for the QCheck.Test.make part)

Pat-Lafon replied:
Hmm, if QCheck has a tuple combinator than sure, otherwise you may need to construct the tuple manually

EnguinandP commented on bin/main.ml:
```diff
@@ -19,18 +19,24 @@ let is_unique l =
-let precondition_frequency_sort = precondition_frequency (fun _ l -> is_sorted l) "is_sorted"
-let precondition_frequency_dup = precondition_frequency (fun _ l -> is_duplicate l) "is_duplicate"
-let precondition_frequency_unique = precondition_frequency (fun _ l -> is_unique l) "is_unique"
+   (gen_type) prop)
+
+let assume_prop prop l = QCheck.assume (prop l); func l
```
Can you explain more what you want this to look like?

Pat-Lafon replied:
I don't think it is necessary(though you can for sure prove me wrong), to pull this(`assume` and our dummy `func l`) out of the above builder.

Basically `gen_type` is of the form `unit -> 'a` and then the `prop` argument should be of the form `'a -> bool`

EnguinandP replied:
Yeah, I had done that because we had `pair (int) (gent_type)` and then `prop` had to be of form `(‘a * ’a) -> bool. However, not all prop had needed a size and list, and you suggested that I remove the pair part. I was trying to come up a solution to that but I think it needs more work

Pat-Lafon replied:
Note that in the sized_list case, `'a = (int, int list)` but for some of the other test cases, it might be something else like `'a = (int, int, int)` or `'a = int tree`

Pat-Lafon reviewed (changes requested):
Ok I've lost a little track of what exactly are outstanding issues or not. 2 things that come to mind are slightly adjusting the signatures of the generators so that will start to align with what I work with and slightly modifying how we pass props to the framework to be less boiler-platey.

(some of the things it lists underneath might be outdated since i listed but didn't request them? Did you end up seeing them??)

EnguinandP commented:
> Ok I've lost a little track of what exactly are outstanding issues or not. 2 things that come to mind are slightly adjusting the signatures of the generators so that will start to align with what I work with and slightly modifying how we pass props to the framework to be less boiler-platey.
> 
> (some of the things it lists underneath might be outdated since i listed but didn't request them? Did you end up seeing them??)

Yeah, I don’t think I saw these… 
But I keep track of the issues usually.

Pat-Lafon commented:
Well hopefully we are synced back up a little. In general, you can always poke me for more clarification about something or if you think I've been oddly silent

Francille committed:
Question about combinators signatures

EnguinandP commented on bin/combinators.ml:
```diff
@@ -2,15 +2,23 @@ let int_gen () = QCheck.Gen.int (QCheck_runner.random_state ())
+let rec int_list_gen' s =
+  if s <= 0 then
+    []
+  else
```
I'm looking at modifying the combinator signatures to look like cobb synthesized generators, which takes an int size (or other parameters). However, to turn them into a' arbitrary using `let arb_builder f = QCheck.make (fun _ -> f ())`, there's no way to pass a size without hardcoding the size in. A solution I found is to use `size : unit -> int` where `arb_builder` is called like `let int_list = arb_builder (int_list_gen nat_gen)`. Knowing this, it seems like there would need to be a wrapper function for the cobb synthesized generators. Is this right or am I missing something?

EnguinandP replied:
Ok, wait I was wrong, I think there's way to pass in a parameter at a later stage.

EnguinandP replied:
Well, regardless, I think that it would make more sense to have wrapper functions for the synthesized generators than the other implementation. For example, `let rec duplicate_list_gen (s : int) (x : int) : int list =`, would be wrapped that in a function with the signature `(unit -> int) -> unit -> int list` where the first argument would be called with my `int_gen`. So the implementation would just look like the first `int_list_gen` in this file with `aux` replaced by `rec duplicate_list_gen`

Pat-Lafon replied:
Not fully clear, if you need to make a wrapper that says `fun () -> let s = nat_gen () in (s, int_list_gen s)` then that is fine. You shouldn't need multiple unit values in args, a bit of an overreaction I think? You just need atleast one argument to get processed to avoid the previous bug

EnguinandP replied:
Ok, I don't think there's any more changes I need to make in the cobb_pbt functions then. How do you think is the best way for getting the cobb synthesized generators in? The only way I can imagine is to manually copy paste them. Is there a module I should look into instead?

Edit: Nevermind, I commented that on the bus and did not remember that the generators were in .ml files. That means I can probably just import them as modules and use them.

Pat-Lafon replied:
Hmm, I'm actually not sure if that will work... the files are .ml but they are weird. What you might need is a simple python script that reads in those files, maybe adds some imports or whatever other boiler plate you need, and then go from there. For sure in a separate directory

EnguinandP replied:
I'm currently trying it with directly using the .ml files. I'm just struggling with dune rn, but it seems fine otherwise?

Pat-Lafon replied:
Cool, it's fine to just hard code the dune file or whatever you need to get something working

EnguinandP replied:
Ok, yeah, I got it to work. I'll push so you can see the example I tried. Look in the new `arbitrary_builder.ml` file I made. 

(the cobb file is just chilling in bin right now, but I will be moving it into a subdirectory)

EnguinandP replied:
Oh wait, let me try to open a new PR for this then.

Francille committed:
Qcheck_runner/my commbinators should be good

Worked on making cli better, but arg.parse is interfering with qcheck runner.

Pat-Lafon commented on bin/combinators.ml:
```diff
@@ -4,22 +4,14 @@ let bool_gen () = QCheck.Gen.bool (QCheck_runner.random_state ())
       int_gen () :: aux (s - 1) in
-  aux s
-
-(* looks like cobb synthesized generators *)
-let rec int_list_gen' s =
-  if s <= 0 then
-    []
-  else
-    int_gen () :: int_list_gen' (s - 1)
-
+  aux size
```
Honestly I liked the other way better. Instead what you might do is make int_list_gen `rec` and then you won't need this intermediate `aux` function.

Thoughts?

EnguinandP replied:
Exactly what I'm thinking!

Side note, how are you able to comment on multiple lines? Everytime it looks like everything is highlighted, it only ends up including some of it

Pat-Lafon replied:
Not sure, here I did a click and drag to highlight the lines I wanted to comment on

EnguinandP replied:
Yeahh that's what I've been doing too

Pat-Lafon commented on bin/main.ml:
```diff
@@ -24,16 +24,17 @@ let precondition_frequency prop name gen_type =
-let precondition_frequency_unique = precondition_frequency (fun _ l -> is_unique l) "is_unique" default_gen
+let precondition_frequency_size = precondition_frequency ((fun prop (n, l) -> QCheck.assume (prop n l); func l ) is_sized) 
+  "is_sized" (QCheck.pair (QCheck.int) (default_gen))
+let precondition_frequency_sort = precondition_frequency is_sorted "is_sorted" default_gen
```
Great! Now lets use int_list explicitly because eventually we will want to have tree benchmarks as well

Pat-Lafon replied:
Moved into it's own issue

Pat-Lafon commented on bin/combinators.ml:
```diff
@@ -1,59 +1,83 @@
+        n :: aux n (s - 1)
+      else 
+        aux prev (s - 1) in
+  aux 0 size
```
`aux 0 size` over `aux (int_gen ()) size` or some equivalent?

EnguinandP replied:
Yeah, let me get that

Francille committed:
Fixed int_list_sorted_gen

Pat-Lafon commented:
Hmm, I think we've started moving on and I'm not sure that the remaining issues are that urgent so I split a few things off into low priority issues and I'll merge this... We can always mess more with this as it comes up
